### PR TITLE
Move to C API

### LINKED ISSUE #22: drop the dependency of cmake?
Currently, `MPI.jl` requires cmake to prepare the shared library. Shall we move to a more Julia way by using BinDeps? I believe most of the stuffes in cmake can be done by BinDeps.

### LINKED ISSUE #59: Binding to the C instead of Fortran interface
Given that Julia is heavily tied to Clang and supports `ccall`, and there is no Clang Fortran compiler, it would clearly be beneficial to switch to binding MPI.jl to MPI's C interface instead of its Fortran interface. The main hurdle is that different MPI implementations define the MPI types in different manners, though one can achieve a high degree of portability with that assumption that {`MPI_Comm`,`MPI_Group`,`MPI_Datatype`,etc.} are either `int` or `void*`, so that one can simply perform checks such as `sizeof(MPI_Comm) == sizeof(int)` to decide on the appropriate datatype.

I would be happy to help with this port if it would be of interest; I'm running into substantial problems with `MPI.jl` on Yosemite.

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #59.

Works around the issues encountered last time (#169) by generating a small C
program which in turn generates the necessary constants, types and
offsets. To avoid the problem of link-time definitions, it uses the
Fortran constants for all handles, which are converted to the C variants
at `MPI_Init`.

This approach has several benefits:
 - No need for CMake (fixes #22) or a Fortran compiler
 - No implicit use of gfortran ABI for strings
 - Simpler to extend, as the new functions only need to be added in one place.

I've also taken the liberty splitting the source into several files
based on functionality, which should make it easier to find things.

### CODE REVIEW

**Comment by simonbyrne** [comment]
The main downside is that 32-bit Windows now fails altogether. I'm not sure what exactly is wrong, but:
```
function Get_version()
    major = Ref{Cint}()
    minor = Ref{Cint}()
    ccall((:MPI_Get_version, libmpi), Cint,
          (Ptr{Cint}, Ptr{Cint}), major, minor)
   return (major[], minor[])
end
```
gives nonsensical results, so it could be an ABI issue?

**Comment by simonbyrne** [comment]
ah, it was a calling convention issue. That also explains #246.

**Comment by andreasnoack** [comment]
Looks good to me. Would be good if @stevengj could take a look, though.

**Comment by lcw** [comment]
I can’t believe this finally got moved to the c api. Thanks @simonbyrne!

**Comment by vchuravy** [comment]
Only managed to test this now on Summit. Sadly it fails with:

```
 22 �[34m~T~L Error: Error building `MPI`:
 23 �[34m~T~B ERROR: LoadError: No MPI library found
```

**Comment by simonbyrne** [comment]
what does the log file say?

File: deps/build.jl [hunk 1]
```diff
@@ -1,285 +1,78 @@
...
-    "MPI_WTICK",
-    "MPI_WTIME",
-    ]
-
-MPI_C_constants = [
-    "MPI_IN_PLACE",
-]
-MPI_C_types = ["Comm", "Info", "Win"]
-
+const libmpi = find_library(Sys.iswindows() ? "msmpi.dll" : "libmpi",
```
Comment (vchuravy) [inline deps/build.jl]: `find_library` checks the provided paths, `DL_LOAD_PATH`, and various system paths. On supercomputers -- as an example see Summit below -- I have rarely seen `DL_LOAD_PATH` being set.
We need a better error message that informs the user can set `JULIA_MPI_PATH`. Alternatively `which mpiexec` might work

```
whatis("Name : spectrum-mpi")
whatis("Version : 10.3.0.0-20190419")
whatis("Short description : IBM MPI implementation from Spectrum MPI.")
help([[IBM MPI implementation from Spectrum MPI.]])
family("mpi")
prepend_path("MODULEPATH","/autofs/nccs-svm1_sw/summit/modulefiles/site/linux-rhel7-ppc64le/spectrum-mpi/10.3.0.0-20190419-4um5hjo/gcc/6.4.0")
setenv("LMOD_MPI_NAME","spectrum-mpi")
setenv("LMOD_MPI_VERSION","10.3.0.0-20190419-4um5hjo")
prepend_path("PATH","/autofs/nccs-svm1_sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-6.4.0/spectrum-mpi-10.3.0.0-20190419-4um5hjogm3tepg4xe23hrptlrs2y7ez6/bin")
prepend_path("MANPATH","/autofs/nccs-svm1_sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-6.4.0/spectrum-mpi-10.3.0.0-20190419-4um5hjogm3tepg4xe23hrptlrs2y7ez6/share/man")
prepend_path("LD_LIBRARY_PATH","/autofs/nccs-svm1_sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-6.4.0/spectrum-mpi-10.3.0.0-20190419-4um5hjogm3tepg4xe23hrptlrs2y7ez6/lib")
prepend_path("LIBRARY_PATH","/autofs/nccs-svm1_sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-6.4.0/spectrum-mpi-10.3.0.0-20190419-4um5hjogm3tepg4xe23hrptlrs2y7ez6/lib")
prepend_path("CPATH","/autofs/nccs-svm1_sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-6.4.0/spectrum-mpi-10.3.0.0-20190419-4um5hjogm3tepg4xe23hrptlrs2y7ez6/include")
prepend_path("CMAKE_PREFIX_PATH","/autofs/nccs-svm1_sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-6.4.0/spectrum-mpi-10.3.0.0-20190419-4um5hjogm3tepg4xe23hrptlrs2y7ez6/")
setenv("OMPI_DIR","/autofs/nccs-svm1_sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-6.4.0/spectrum-mpi-10.3.0.0-20190419-4um5hjogm3tepg4xe23hrptlrs2y7ez6")
setenv("MPI_ROOT","/autofs/nccs-svm1_sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-6.4.0/spectrum-mpi-10.3.0.0-20190419-4um5hjogm3tepg4xe23hrptlrs2y7ez6")
setenv("OPAL_PREFIX","/autofs/nccs-svm1_sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-6.4.0/spectrum-mpi-10.3.0.0-20190419-4um5hjogm3tepg4xe23hrptlrs2y7ez6")
setenv("OPAL_LIBDIR","/autofs/nccs-svm1_sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-6.4.0/spectrum-mpi-10.3.0.0-20190419-4um5hjogm3tepg4xe23hrptlrs2y7ez6/lib")
setenv("OMPI_CC","gcc")
setenv("OMPI_CXX","g++")
setenv("OMPI_FC","gfortran")
setenv("OLCF_SPECTRUM_MPI_ROOT","/autofs/nccs-svm1_sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-6.4.0/spectrum-mpi-10.3.0.0-20190419-4um5hjogm3tepg4xe23hrptlrs2y7ez6")
setenv("PAMI_IBV_ENABLE_OOO_AR","1")
setenv("PAMI_IBV_QP_SERVICE_LEVEL","8")
```

Comment (simonbyrne) [reply deps/build.jl]: if you set `JULIA_MPI_PATH` does it work?

Comment (simonbyrne) [reply deps/build.jl]: or do you need to `module load XXmpiXX`?

**Pull request merged**